interfaces/builtin/dotfiles.go

```diff
@@ -0,0 +1,158 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const dotfilesSummary = `allows access to files or directories`
+
+const dotfilesBaseDeclarationSlots = `
+  dotfiles:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const dotfilesConnectedPlugAppArmor = `
+# Description: Can access specific files or directories.
+# This is restricted because it gives file access to arbitrary locations.
+`
+
+type dotfilesInterface struct {
+	commonInterface
+}
+
+func validatePaths(attrName string, paths []interface{}) error {
+	for _, npp := range paths {
+		np, ok := npp.(string)
+		if !ok {
+			return fmt.Errorf("%q must be a list of strings", attrName)
+		}
+		// filepath.Clean() will remove trailing "/" but we allow this
+		// to differentiate between files or dirs
+		last := len(np)
+		if np[last-1] == '/' {
+			np = np[:last-1]
+		}
+		if !strings.HasPrefix(np, "/") && !strings.HasPrefix(np, "$HOME/") {
+			return fmt.Errorf(`%q must start with "/" or "$HOME"`, np)
+		}
+		p := filepath.Clean(np)
+		if p != np {
+			return fmt.Errorf("%q must be clean", np)
+		}
+		if strings.Contains(p, "..") {
+			return fmt.Errorf(`%q contains invalid ".."`, p)
+		}
+		if strings.Contains(p, "~") {
+			return fmt.Errorf(`%q contains invalid "~"`, p)
+		}
+		if err := apparmor.ValidateFreeFromAARE(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (iface *dotfilesInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	hasValidAttr := false
+	for _, att := range []string{"read", "write"} {
+		if _, ok := plug.Attrs[att]; !ok {
+			continue
+		}
+		il, ok := plug.Attrs[att].([]interface{})
+		if !ok {
+			return fmt.Errorf("cannot add dotfiles plug: %q must be a list of strings", att)
+		}
+		if err := validatePaths(att, il); err != nil {
+			return fmt.Errorf("cannot add dotfiles plug: %s", err)
+		}
+		hasValidAttr = true
+	}
+	if !hasValidAttr {
+		return fmt.Errorf(`cannot add dotfiles plug: needs valid "read" or "write" attribute`)
+	}
+
+	return nil
+}
+
+func formatPath(ip interface{}) (string, error) {
+	p, ok := ip.(string)
+	if !ok {
+		return "", fmt.Errorf("%[1]v (%[1]T) is not a string", ip)
+	}
+	prefix := ""
+	if strings.Count(p, "$HOME") > 0 {
+		p = strings.Replace(p, "$HOME", "@${HOME}", 1)
+		prefix = "owner "
+	}
+	return prefix + filepath.Clean(p), nil
```

Comment:
We are going to want to end up with: `owner "@{HOME}/foo" rk,`, but AFAICS, this will end up with `owner @{HOME}/foo rk,` (ie, the quotes are dropped).

Reply:
I added quotes around all user supplied paths now.